python/faery/events_stream.py

```diff
@@ -439,22 +509,61 @@ def apply(
     ) -> "FiniteRegularEventsStream":
         return filter_class(self, *args, **kwargs)  # type: ignore
 
-    def to_array(self) -> numpy.ndarray:
-        return numpy.concatenate(list(self))
+    def to_array(
+        self, on_progress: typing.Callable[[OutputState], None] = lambda _: None
+    ) -> numpy.ndarray:
+        events_buffers = []
+        state_manager = events_stream_state.StateManager(
+            stream=self, on_progress=on_progress
+        )
+        state_manager.start()
+        for events in self:
+            events_buffers.append(events)
+            state_manager.commit(events=events)
+        if len(events_buffers) == 0:
+            result = numpy.array([], dtype=EVENTS_DTYPE)
+        else:
+            result = numpy.concatenate(events_buffers, dtype=EVENTS_DTYPE)
+        state_manager.end()
+        return result
 
     def envelope(
         self,
         decay: enums.Decay,
         tau: timestamp.Time,
     ) -> frame_stream.FiniteRegularFloat64FrameStream: ...
 
+    def to_kinectograph(
```

Comment:
I like that the work is done in a class - that we can then "pull into" the streaming API. That makes the code much less messy